File: leetcode/e/Unique Word Abbreviation.java
/*
 * An abbreviation of a word follows the form <first letter><number><last letter>. Below are some examples of word abbreviations:

a) it                      --> it    (no abbreviation)

     1
b) d|o|g                   --> d1g

              1    1  1
     1---5----0----5--8
c) i|nternationalizatio|n  --> i18n

              1
     1---5----0
d) l|ocalizatio|n          --> l10n
Assume you have a dictionary and given a word, find whether its abbreviation is unique in the dictionary. A word's abbreviation is unique if no other word from the dictionary has the same abbreviation.

Example: 
Given dictionary = [ "deer", "door", "cake", "card" ]

isUnique("dear") -> false
isUnique("cart") -> true
isUnique("cane") -> false
isUnique("make") -> true
* /
public class ValidWordAbbr {

    Map<String, Set<String>> abbrToWords = new HashMap(); 

    public ValidWordAbbr(String[] dictionary) {
        for (String word : dictionary) {
            String abbr = abbr(word);
            Set<String> words = abbrToWords.get(abbr);
            if (words == null) {
                words = new HashSet<String>();
                abbrToWords.put(abbr, words);
            }
            words.add(word);
        }
    }
    

    public boolean isUnique(String word) {
        Set<String> set = abbrToWords.get(abbr(word));
        if (set == null) return true; // ERROR: no abbr meas true
        if (set.size() != 1 || !set.contains(word)) return false; // ERROR: even if abbr exist and size = 1, need to check if query word is the word in dict
        return true;
    }
    
    private String abbr(String word) {
        if (word.length() <= 2) return word;
        int len = word.length() - 2;
        return "" + word.charAt(0) + len + word.charAt(word.length() -1);
    }
}


// Your ValidWordAbbr object will be instantiated and called as such:
// ValidWordAbbr vwa = new ValidWordAbbr(dictionary);
// vwa.isUnique("Word");
// vwa.isUnique("anotherWord");

File: lc_plusplus/Unique Word Abbreviation.cc
// An abbreviation of a word follows the form <first letter><number><last letter>. Below are some examples of word abbreviations:
//
// a) it                      --> it    (no abbreviation)
//
//      1
// b) d|o|g                   --> d1g
//
//               1    1  1
//      1---5----0----5--8
// c) i|nternationalizatio|n  --> i18n
//
//               1
//      1---5----0
// d) l|ocalizatio|n          --> l10n
// Assume you have a dictionary and given a word, find whether its abbreviation is unique in the dictionary. A word's abbreviation is unique if no other word from the dictionary has the same abbreviation.
//
// Example:
// Given dictionary = [ "deer", "door", "cake", "card" ]
//
// isUnique("dear") -> false
// isUnique("cart") -> true
// isUnique("cane") -> false
// isUnique("make") -> true


class ValidWordAbbr {
    unordered_map<string, unordered_set<string>> abbrToWords;
public:
    ValidWordAbbr(vector<string> dictionary) {
        for (auto& w : dictionary) {
          abbrToWords[abbr(w)].insert(w);
        }
    }

    bool isUnique(string word) {
        auto abbr = this->abbr(word);
        return (abbrToWords.count(abbr) == 0 || (abbrToWords[abbr].size() == 1 && abbrToWords[abbr].count(word) == 1));
    }

    string abbr(string w) {
      string abbr;
      if (w.size() < 3) abbr = w;
      else {
        abbr += w[0];
        abbr += to_string(w.size() - 2);
        abbr += w[w.size() - 1];
      }
      return abbr;
    }
};

/**
 * Your ValidWordAbbr object will be instantiated and called as such:
 * ValidWordAbbr obj = new ValidWordAbbr(dictionary);
 * bool param_1 = obj.isUnique(word);
 */
